src/VecSim/algorithms/svs/svs_tiered.h

```diff
@@ -0,0 +1,726 @@
+        std::shared_lock<std::shared_mutex> main_lock(this->mainIndexGuard);
+        return this->frontendIndex->indexSize() + this->backendIndex->indexSize();
+    }
+
+    size_t indexLabelCount() const override {
+        std::shared_lock<std::shared_mutex> flat_lock(this->flatIndexGuard);
+        std::shared_lock<std::shared_mutex> main_lock(this->mainIndexGuard);
+        return this->frontendIndex->indexLabelCount() + this->backendIndex->indexLabelCount();
+    }
+    size_t indexCapacity() const override {
+        std::shared_lock<std::shared_mutex> flat_lock(this->flatIndexGuard);
+        std::shared_lock<std::shared_mutex> main_lock(this->mainIndexGuard);
+        return this->frontendIndex->indexCapacity() + this->backendIndex->indexCapacity();
+    }
+
+    double getDistanceFrom_Unsafe(labelType label, const void *blob) const override {
+        double flat_dist = std::numeric_limits<double>::quiet_NaN();
+        {
+            std::shared_lock<decltype(this->flatIndexGuard)> lock(this->flatIndexGuard);
+            flat_dist = this->frontendIndex->getDistanceFrom_Unsafe(label, blob);
+        }
+        if (!std::isnan(flat_dist)) {
+            return flat_dist;
+        } else {
+            std::shared_lock<decltype(this->mainIndexGuard)> lock(this->mainIndexGuard);
+            return this->backendIndex->getDistanceFrom_Unsafe(label, blob);
+        }
+    }
+
+    VecSimIndexDebugInfo debugInfo() const override {
+        auto info = Base::debugInfo();
+        return info;
+    }
+
+    VecSimIndexBasicInfo basicInfo() const override {
+        VecSimIndexBasicInfo info = this->backendIndex->getBasicInfo();
+        info.blockSize = info.blockSize;
+        info.isTiered = true;
+        info.algo = VecSimAlgo_SVS;
+        return info;
+    }
+
+    VecSimDebugInfoIterator *debugInfoIterator() const override {
+        //  Get the base tiered fields.
+        auto *infoIterator = Base::debugInfoIterator();
+        return infoIterator;
+    }
+
+    VecSimBatchIterator *newBatchIterator(const void *queryBlob,
+                                          VecSimQueryParams *queryParams) const override {
+        size_t blobSize = this->backendIndex->getDim() * sizeof(DataType);
+        void *queryBlobCopy = this->allocator->allocate(blobSize);
+        memcpy(queryBlobCopy, queryBlob, blobSize);
+        return new (this->allocator)
+            TieredSVS_BatchIterator(queryBlobCopy, this, queryParams, this->allocator);
+    }
+
+    void setLastSearchMode(VecSearchMode mode) override {
+        return this->backendIndex->setLastSearchMode(mode);
+    }
+
+    void runGC() override {
+        // Run no more than pendingSwapJobsThreshold value jobs.
+        TIERED_LOG(VecSimCommonStrings::LOG_VERBOSE_STRING,
+                   "running asynchronous GC for tiered SVS index");
+        if (!indexUpdateScheduled.test_and_set()) {
+            updateSVSIndexWrapper(this, 1);
```

Comment:
why are we calling this here?
what cleanup operation does it do? is it going to trigger indexing operations as well?